File: src/commands/brokers.ts
import { getRunEnv, RUN_ENV, invokeTauri, invokeWeb, logger } from "@/adapters";

export interface BrokerConnectionDto {
  id: string;
  accountId: string;
  provider: string;
  createdAt: string;
  updatedAt: string;
}

export const getAuthUrl = async (provider: string, profileId?: string): Promise<string> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_auth_url", { provider, profileId });
      case RUN_ENV.WEB:
        return invokeWeb("get_auth_url", { provider, profileId });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error getting auth URL.");
    throw error;
  }
};

export const handleOauthCallback = async (accountId: string, provider: string, profileId?: string): Promise<BrokerConnectionDto> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("handle_oauth_callback", { accountId, provider, profileId });
      case RUN_ENV.WEB:
        return invokeWeb("handle_oauth_callback", { accountId, provider, profileId });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error handling OAuth callback.");
    throw error;
  }
};

export const getBrokerConnection = async (accountId: string, profileId?: string): Promise<BrokerConnectionDto | null> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_broker_connection", { accountId, profileId });
      case RUN_ENV.WEB:
        return invokeWeb("get_broker_connection", { accountId, profileId });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error getting broker connection.");
    throw error;
  }
};

export const fetchBrokerActivities = async (accountId: string, profileId?: string): Promise<any[]> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("fetch_broker_activities", { accountId, profileId });
      case RUN_ENV.WEB:
        return invokeWeb("fetch_broker_activities", { accountId, profileId });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error fetching broker activities.");
    throw error;
  }
};

File: src-server/src/api/brokers.rs
use axum::{
    extract::{Path, State, Query},
    routing::{get, post},
    Json, Router,
};
use std::collections::HashMap;
use std::sync::Arc;
use investwise_core::brokers::models::BrokerConnectionDto;
use investwise_core::activities::ActivityImport;
use crate::main_lib::AppState;

pub fn brokers_routes() -> Router<Arc<AppState>> {
    Router::new()
        .route("/auth-url/{provider}", get(get_auth_url))
        .route("/callback/{account_id}/{provider}", post(handle_oauth_callback))
        .route("/connections/{account_id}", get(get_broker_connection))
        .route("/broker-activities/{account_id}", post(fetch_broker_activities))
}

async fn get_auth_url(
    State(state): State<Arc<AppState>>,
    Path(provider): Path<String>,
    Query(query): Query<HashMap<String, String>>,
) -> Result<Json<String>, String> {
    let profile_id = query.get("profile_id").map(|s| s.as_str()).unwrap_or_default();
    state.broker_service.get_auth_url(&provider, profile_id).await.map(Json).map_err(|e| e.to_string())
}

async fn handle_oauth_callback(
    State(state): State<Arc<AppState>>,
    Path((account_id, provider)): Path<(String, String)>,
    Query(query): Query<HashMap<String, String>>,
) -> Result<Json<BrokerConnectionDto>, String> {
    let profile_id = query.get("profile_id").map(|s| s.as_str()).unwrap_or_default();
    state.broker_service.handle_oauth_callback(&account_id, &provider, profile_id)
        .await
        .map(Json)
        .map_err(|e| e.to_string())
}

async fn get_broker_connection(
    State(state): State<Arc<AppState>>,
    Path(account_id): Path<String>,
    Query(query): Query<HashMap<String, String>>,
) -> Result<Json<Option<BrokerConnectionDto>>, String> {
    let profile_id = query.get("profile_id").map(|s| s.as_str()).unwrap_or_default();
    state.broker_service.get_connections_by_account(&account_id, profile_id)
        .map(Json)
        .map_err(|e| e.to_string())
}

async fn fetch_broker_activities(
    State(state): State<Arc<AppState>>,
    Path(account_id): Path<String>,
    Query(query): Query<HashMap<String, String>>,
) -> Result<Json<Vec<ActivityImport>>, String> {
    let profile_id = query.get("profile_id").map(|s| s.as_str()).unwrap_or_default();
    state.broker_service.fetch_broker_activities(&account_id, profile_id)
        .await
        .map(Json)
        .map_err(|e| e.to_string())
}
